rng/philox/philox.hpp

```diff
@@ -0,0 +1,269 @@
+    using promotion_types = std::tuple<std::uint16_t, std::uint32_t, std::uint64_t, __uint128_t>;
+public:
+    static consteval std::size_t get_log_index(std::size_t val)
+    {
+        val = (val - 1) / CHAR_BIT + 1;
+        auto width = std::bit_width(val);
+
+        std::size_t subtrahend = static_cast<std::size_t>(std::has_single_bit(val));
+
+        return width - subtrahend;
+    }
+private:
+
+    using counter_type = std::tuple_element_t<get_log_index(w), uint_types>;
+    using promotion_type = std::tuple_element_t<get_log_index(w), promotion_types>;
+
+    static constexpr counter_type counter_mask = ~counter_type(0) >> (sizeof(counter_type) * CHAR_BIT - w);
+    static constexpr result_type result_mask = ~result_type(0) >> (sizeof(result_type) * CHAR_BIT - w);
+
+
+private: // functions
+
+    static std::pair<counter_type, counter_type> mulhilo(result_type a, result_type b)
+    {
+        constexpr std::size_t shift = std::numeric_limits<promotion_type>::digits - w;
+        promotion_type promoted_a = a;
+        promotion_type promoted_b = b;
+        promotion_type result = promoted_a * promoted_b;
+        counter_type mulhi = result >> shift;
+        counter_type mullo = (result << shift) >> shift;
+        return {mulhi, mullo};
+    }
+
+    static std::array<result_type, n> philox_generate(std::array<result_type, array_size> keys, std::array<counter_type, n> x)
+    {
+        for (std::size_t q = 0; q != r; ++q)
+        {
+            static constexpr std::array<std::array<std::size_t, 16>, 4> permute_indices
+            {{
+                {0, 1},
+                {2, 1, 0, 3},
+                {0, 5, 2, 7, 6, 3, 4, 1},
+                {2, 1, 4, 9, 6, 15, 0, 3, 10, 13, 12, 11, 14, 7, 8, 5}
+            }};
+            auto row = get_log_index(n * CHAR_BIT) - 1;
+            std::array<counter_type, n> v = [&x, row]<std::size_t... Is>(std::index_sequence<Is...>) {
+                return std::array<counter_type, n>{x[permute_indices[row][Is]]...};
+            }(std::make_index_sequence<n>{});
+
+            for (std::size_t k = 0; k < array_size; ++k)
+            {
+                auto [mulhi, mullo] = mulhilo(v[2 * k], multipliers[k]);
+                x[2 * k + 1] = mullo;
+                x[2 * k] = mulhi ^ keys[k] ^ v[2 * k + 1];
+                keys[k] = (keys[k] + round_consts[k]) & result_mask;
+            }
+        }
+        return [&x]<std::size_t... Is>(std::index_sequence<Is...>) {
+            return std::array<result_type, n>{x[Is]...};
+        }(std::make_index_sequence<n>{});
+    }
+
+    void increment_counter()
+    {
+        for (auto& elem : x)
+        {
+            ++elem;
+            elem &= counter_mask;
+            if (elem != 0)
+            {
+                break;
+            }
+        }
+    }
+
+    void increment_counter(unsigned long long z)
+    {
+        for(std::size_t j = 0; j < n; j++) {
+            z += (unsigned long long)x[j];
```

Comment:
z = -1; // max unsigned long long
x[0] = 1;
z += x[0]; // -> z=0; overflow, skipping zero elements!

Reply:
Seems like it works now. I don't like current implementation from C++ perspective (as we discussed) but we decided to apply the quickest fix to have a correct implementation and then improve it if necessary